stlab/future.hpp

```diff
@@ -814,23 +814,24 @@ struct when_all_shared {
     std::tuple<boost::optional<Ts>...>  _args;
     future<void>                        _holds[sizeof...(Ts)] {};
     std::atomic_size_t                  _remaining {sizeof...(Ts)};
-    std::atomic_bool                    _error_happend{ false };
-    std::mutex                          _errormutex;
+    std::atomic_bool                    _error_happened{ false };
     boost::optional<std::exception_ptr> _error;
     packaged_task<>                     _f;
 
-    void done() { if (--_remaining == 0 && !_error_happend) _f(); }
+    void done() { if (!_error_happened && --_remaining == 0) _f(); }
```

Comment:
Why was the ordering changed (at a minimum this needs a comment if the ordering is required). This will now typically have to perform two atomic operations instead of one. I also think there is a race - if an error has not happened, then we decrement but an error can occur between the check and the decrement. With the original order, if we decrement and hit zero, we know a failure cannot occur after so we can then check to see if one has happened.